src/map_func.h

```diff
@@ -427,6 +407,112 @@ debug_inline static uint TileY(TileIndex tile)
 	return tile.base() >> Map::LogX();
 }
 
+/**
+ * An offset value between two tiles.
+ *
+ * This value is used for the difference between two tiles. It can be added to
+ * a TileIndex to get the resulting TileIndex of the start tile applied with
+ * this saved difference.
+ *
+ * When being built with debugging, the internal structure is different to be
+ * able to perform stringent overflow checks on the X and Y axes. However,
+ * that structure has negative performance impacts not warranted for releases.
+ *
+ * @see TileDiffXY(int, int)
+ */
+struct TileIndexDiff {
+public:
+	friend inline TileIndexDiff TileDiffXY(int x, int y);
+	friend inline TileIndex operator-(const TileIndex &tile, const TileIndexDiff &offset);
+	friend inline TileIndex operator+(const TileIndex &tile, const TileIndexDiff &offset);
+
+	debug_inline constexpr TileIndexDiff operator*(int amount) const
+	{
+#ifdef _DEBUG
+		return TileIndexDiff(this->x * amount, this->y * amount);
+#else
+		return TileIndexDiff(this->offset * amount);
+#endif /* _DEBUG */
+	}
+
+	debug_inline constexpr TileIndexDiff& operator*=(int amount)
+	{
+#ifdef _DEBUG
+		this->x *= amount;
+		this->y *= amount;
+#else
+		this->offset *= amount;
+#endif /* _DEBUG */
```

Comment:
I think maybe the `_DEBUG` path should not exclude the normal path, and should be able to test that the result is the same. Otherwise we have to assume, which isn't good.